Support non-1 indices and fix type problems in DFT (fixes #17896)

timholy opened the pull request:
This supports FFT on non-1 arrays, by introducing `circcopy!`:

``` jl
julia> using OffsetArrays

julia> a = OffsetArray(1:7, -1:5)
OffsetArrays.OffsetArray{Int64,1,UnitRange{Int64}} with indices -1:5:
 1
 2
 3
 4
 5
 6
 7

julia> a1 = Array{Int}(7);

julia> circcopy!(a1, a)
7-element Array{Int64,1}:
 3
 4
 5
 6
 7
 1
 2

julia> a[1:5] == a1[1:5]
true
```

As a consequence, index information gets preserved in the phase of the FFT. I tried an alternative implementation, multiplying by phases _post hoc_, but in my hands that was slower.

I also took the opportunity to fix (and test for) #17896.

CC @stevengj.

StefanKarpinski commented:
I'm glad you're pushing this forward, @timholy! Any change that shouldn't break code and isn't too risky should probably be marked for backporting to release-0.5. Does this qualify?

StefanKarpinski commented:
Minus the export, I should add.

timholy commented:
Yes, it qualifies. Thanks for the reminder re the label.

> I'm glad you're pushing this forward

Nothing like having a new real-world application to put it through its paces...the application I used while developing non-1 indices in the first place was based on old code where I faked unconventional indices, but perforce that couldn't stretch the limits like it's possible to do now. Once some stuff works, you quickly discover how much more you want to get working.

tkelman commented on base/multidimensional.jl:
```diff
@@ -662,6 +662,66 @@ function _circshift!(dest, rdest, src, rsrc, inds, shiftamt)
 
+# circcopy!
+"""
+    circcopy!(dest, src)
```
for 0.6 this should go in the rst manual if exported

tkelman commented on base/multidimensional.jl:
```diff
@@ -662,6 +662,66 @@ function _circshift!(dest, rdest, src, rsrc, inds, shiftamt)
+    rs1, rs2 = first(inds1):ssf, ssf+1:last(inds1)
+    tindsd, tindss = tail(indsdest), tail(indssrc)
+    _circcopy!(dest, (rdest..., rd1), tindsd, src, (rsrc..., rs2), tindss)
+    _circcopy!(dest, (rdest..., rd2), tindsd, src, (rsrc..., rs1), tindss)
```
this could use some comments

timholy committed:
Support non-1 indices and fix type problems in DFT (fixes #17896)

timholy committed:
Export circcopy! and add it to the RST manual

timholy commented:
I split the `export` and RST manual addition into a separate commit, so that when backporting to release-0.5 you can just use the first commit.

I also added comments as requested.

tkelman commented on doc/stdlib/arrays.rst:
```diff
@@ -616,6 +616,33 @@ Indexing, Assignment, and Concatenation
+        3  7  11  15
+        4  8  12  16
+
+       julia> dest = OffsetArray(Int, (-1,1));
```
this doctest won't pass without some setup code to load the OffsetArray type definition from `test/`, will it?

timholy replied:
Good catch. I'm guessing we can't just say `using OffsetArrays`, for reasons previously discussed. Is there some kind of "hidden" setup possible? Or should I just remove the doctest and make sure this runs in the main tests?

tkelman replied:
I believe there is a way to do a few lines of test setup that don't appear in the rendered docs. It might be worth moving the definition of the type and methods from `test/offsetarray.jl` to `test/TestHelpers.jl`, if the rest of the correctness tests take a little too long to run as part of the doctest setup.

timholy replied:
I just got rid of it, as (1) I was not successful in figuring out hiding, and (2) the printing of the typename just didn't seem to benefit from having TestHelpers in it. I did move the OAs module to TestHelpers.jl, however.

stevengj commented on base/dft.jl:
```diff
@@ -20,22 +20,41 @@ export fft, ifft, bfft, fft!, ifft!, bfft!,
+realfloat{T<:Real}(x::AbstractArray{T}) = copy1(typeof(fftwfloat(one(T))), x)
+
+# copy to a 1-based array, using circular permutation
+function copy1{T}(::Type{T}, x)
```
I don't understand why it would be necessary to make a copy.  FFTW doesn't care how Julia indexes the array, as long as the data is strided.  It should be possible to make it work with any `DenseArray` (i.e. supporting `pointer` and `strides`) without copying.

timholy replied:
Mathematical consistency. As you know better than almost anyone, here's the formula:
```latex
FFT(x)[k] = \sum_{n=1}^N x[n] exp(-2\pi (n-1)(k-1)/N)
```
`x[n]` is to be interpreted with periodic boundary conditions.


That summation range actually means something, and 1 is **not** synonymous with "the first value." If my array is defined over `-1:5`, it's equivalent to it being defined over `1:7` but you need to circularly-permute the values. Nothing else makes any sense.

Incidentally, this makes array-padding---extending the indices in either direction---an absolute dream. No more index gymnastics necessary.

timholy replied:
BTW, I do think we are a bit confused about `DenseArray` vs `StridedArray`: if `DenseArray` means it supports `pointer` and `strides`, what exactly does `StridedArray` mean?

stevengj replied:
You're right, I meant `StridedArray`.

I see your point, you want the indexing to define the "origin" (the phase offset) of the DFT.

stevengj replied:
Note that you could still do it without making a copy by multiplying the output by an appropriate phase.  It's probably more expensive than a copy, though.

timholy replied:
In my tests, it was.

timholy committed:
Move OffsetArray definition to TestHelpers

timholy commented:
I think this can be merged once CI passes.
